### PR TITLE
[Feature] support cast string and json to array type

### LINKED ISSUE #11306: Support cast string to array type
## Feature request

**Is your feature request related to a problem? Please describe.**

Support the following type casting:
```sql

-- string to array<ANY>
select cast('[1,2,3]' as array<int>);
-- [1,2,3]

select cast('[1,2,3]' as array<bigint>);
-- [1,2,3]

select cast('[1,2,3]' as array<string>);
-- ["1","2","3"]

select cast('["a", "b", "c"]' as array<string>);
-- ["a","b","c"]

-- json array to array<ANY>
select cast(parse_json('[{"a":1}, {"a": 2}]')  as array<json>);
-- ['{"a": 1}','{"a": 2}']

select cast(parse_json('[1, 2, 3]')  as array<int>);
-- [1, 2, 3]

select cast(parse_json('["1","2","3"]') as array<string>);
-- ["1","2","3"]

```

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by murphyatwork**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #11306 #9827

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by mofeiatwork** [comment]
@Mergifyio backport branch-2.4

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 1 / 2 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Type.java|1|2|50.00%|[962]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 1 / 2 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Type.java|1|2|50.00%|[962]|

File: be/src/exprs/vectorized/cast_expr.cpp [hunk 1]
```diff
@@ -1423,6 +1446,24 @@ Expr* VectorizedCastExprFactory::from_thrift(const TExprNode& node, bool allow_t
             return new VectorizedCastExpr<TYPE_VARCHAR, TYPE_HLL, false>(node);
         }
     }
+    // Cast string to array<ANY>
+    if (from_type == TYPE_VARCHAR && to_type == TYPE_ARRAY) {
+        TypeDescriptor cast_to = TypeDescriptor::from_thrift(node.type);
+        TExprNode cast;
+        cast.type = cast_to.children[0].to_thrift();
```
Comment (kangkaisen) [inline be/src/exprs/vectorized/cast_expr.cpp:1453]: why not set array element type info in FE

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 1 / 1 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Type.java|1|1|100.00%|[]|

File: be/src/exprs/vectorized/cast_expr.cpp (hunk 1 above)
Comment (mofeiatwork) [reply be/src/exprs/vectorized/cast_expr.cpp:1453]: element type is set into `children` for complex types.

File: be/src/column/column_helper.h [hunk 2]
```diff
@@ -166,20 +165,28 @@ class ColumnHelper {
         }
     }
 
+    // Cast to Nullable
+    static ColumnPtr cast_to_nullable_column(const ColumnPtr& src_column) {
```
Comment (fzhedu) [inline be/src/column/column_helper.h:169]: inline?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 1 / 1 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Type.java|1|1|100.00%|[]|

**Review by kangkaisen — DISMISSED** [review]

File: be/src/column/column_helper.h (hunk 2 above)
Comment (mofeiatwork) [reply be/src/column/column_helper.h:169]: This function is located at header file, so there's no difference of using a `inline` keyword

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run all

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_fe_unittest

File: be/src/exprs/vectorized/cast_expr_array.cpp [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    while (slice.starts_with(" ")) {
+        slice.remove_prefix(1);
+    }
+    while (slice.ends_with(" ")) {
+        slice.remove_suffix(1);
+    }
+    if (slice.starts_with("\"") || slice.starts_with("'")) {
+        slice.remove_prefix(1);
+    }
+    if (slice.ends_with("\"") || slice.ends_with("'")) {
```
Comment (ZiheLiu) [inline be/src/exprs/vectorized/cast_expr_array.cpp:142]: - What's the result for `select cast('["1\']' as array<string>)`? Should it be `["\"1'"]` or `[null]`?
- And what's the result for `select cast('[a]' as array<string>)`? Should it be `["a"]` or `[null]`?

File: be/src/exprs/vectorized/cast_expr.h [hunk 4]
```diff
@@ -9,13 +9,17 @@
...
 namespace starrocks {
 namespace vectorized {
 
 class VectorizedCastExprFactory {
 public:
-    static Expr* from_thrift(const TExprNode& node, bool exception_if_failed = false);
+    static Expr* from_thrift(const TExprNode& node, bool exception_if_failed = false) {
+        return from_thrift(nullptr, node, exception_if_failed);
+    }
+    static Expr* from_thrift(ObjectPool* pool, const TExprNode& node, bool exception_if_failed = false);
```
[the remark this answers is not in the document]
Comment (mofeiatwork) [reply be/src/exprs/vectorized/cast_expr.h:22]: done

File: be/src/exprs/vectorized/cast_expr_array.cpp (hunk 3 above)
Comment (mofeiatwork) [reply be/src/exprs/vectorized/cast_expr_array.cpp:142]: The result should be changed into: 
1. `select cast('["1\']' as array<string>)` => `["\"1\'"]`, a string with escaped quote
2. `select cast('[a]' as array<string>)` => `["a"], which should be correct casted

**Review by ZiheLiu — DISMISSED** [review]

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_fe_unittest

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 1 / 2 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Type.java|1|2|50.00%|[959]|

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 9 / 11 (81.82%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|1|2|50.00%|[441]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|8|9|88.89%|[960]|

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Review by kangkaisen — APPROVED** [review]

**Review by ZiheLiu — APPROVED** [review]

**Pull request merged**